scapy/contrib/iec104.py

```diff
@@ -0,0 +1,587 @@
+          one goes to implement a iec101<-->iec104 gateway or hits such a gateway that acts not standard conform
+          (e.g. by forwarding 101 messages to a 104 network)
+"""
+
+from scapy.compat import orb
+from scapy.contrib.iec104_information_objects import *
+
+from scapy.contrib.iec104_fields import LEThreeBytesField, IEC104SequenceNumber
+from scapy.contrib.iec104_information_objects import IEC104_IO_NAMES, IEC104_IO_WITH_IOA_CLASSES, IEC104_IO_CLASSES
+from scapy.error import warning, Scapy_Exception
+from scapy.fields import ByteField, BitField, ByteEnumField, PacketListField, \
+    BitEnumField, XByteField, FieldLenField, LEShortField
+
+from scapy.layers.inet import TCP
+from scapy.packet import Packet, bind_layers
+
+IEC_104_IANA_PORT = 2404
+
+# direction - from the central station to the substation
+IEC104_CONTROL_DIRECTION = 0
+IEC104_CENTRAL_2_SUB_DIR = IEC104_CONTROL_DIRECTION
+
+# direction - from the substation to the central station
+IEC104_MONITOR_DIRECTION = 1
+IEC104_SUB_2_CENTRAL_DIR = IEC104_MONITOR_DIRECTION
+
+IEC104_DIRECTIONS = {
+    IEC104_MONITOR_DIRECTION: 'monitor direction (sub -> central)',
+    IEC104_CONTROL_DIRECTION: 'control direction (central -> sub)',
+}
+
+# COT - cause of transmission
+IEC104_COT_UNDEFINED = 0
+IEC104_COT_CYC = 1
+IEC104_COT_BACK = 2
+IEC104_COT_SPONT = 3
+IEC104_COT_INIT = 4
+IEC104_COT_REQ = 5
+IEC104_COT_ACT = 6
+IEC104_COT_ACTCON = 7
+IEC104_COT_DEACT = 8
+IEC104_COT_DEACTCON = 9
+IEC104_COT_ACTTERM = 10
+IEC104_COT_RETREM = 11
+IEC104_COT_RETLOC = 12
+IEC104_COT_FILE = 13
+IEC104_COT_RESERVED_14 = 14
+IEC104_COT_RESERVED_15 = 15
+IEC104_COT_RESERVED_16 = 16
+IEC104_COT_RESERVED_17 = 17
+IEC104_COT_RESERVED_18 = 18
+IEC104_COT_RESERVED_19 = 19
+IEC104_COT_INROGEN = 20
+IEC104_COT_INRO1 = 21
+IEC104_COT_INRO2 = 22
+IEC104_COT_INRO3 = 23
+IEC104_COT_INRO4 = 24
+IEC104_COT_INRO5 = 25
+IEC104_COT_INRO6 = 26
+IEC104_COT_INRO7 = 27
+IEC104_COT_INRO8 = 28
+IEC104_COT_INRO9 = 29
+IEC104_COT_INRO10 = 30
+IEC104_COT_INRO11 = 31
+IEC104_COT_INRO12 = 32
+IEC104_COT_INRO13 = 33
+IEC104_COT_INRO14 = 34
+IEC104_COT_INRO15 = 35
+IEC104_COT_INRO16 = 36
+IEC104_COT_REQCOGEN = 37
+IEC104_COT_REQCO1 = 38
+IEC104_COT_REQCO2 = 39
+IEC104_COT_REQCO3 = 40
+IEC104_COT_REQCO4 = 41
+IEC104_COT_RESERVED_42 = 42
+IEC104_COT_RESERVED_43 = 43
+IEC104_COT_UNKNOWN_TYPE_CODE = 44
+IEC104_COT_UNKNOWN_TRANSMIT_REASON = 45
+IEC104_COT_UNKNOWN_COMMON_ADDRESS_OF_ASDU = 46
+IEC104_COT_UNKNOWN_ADDRESS_OF_INFORMATION_OBJECT = 47
+IEC104_COT_PRIVATE_48 = 48
+IEC104_COT_PRIVATE_49 = 49
+IEC104_COT_PRIVATE_50 = 50
+IEC104_COT_PRIVATE_51 = 51
+IEC104_COT_PRIVATE_52 = 52
+IEC104_COT_PRIVATE_53 = 53
+IEC104_COT_PRIVATE_54 = 54
+IEC104_COT_PRIVATE_55 = 55
+IEC104_COT_PRIVATE_56 = 56
+IEC104_COT_PRIVATE_57 = 57
+IEC104_COT_PRIVATE_58 = 58
+IEC104_COT_PRIVATE_59 = 59
+IEC104_COT_PRIVATE_60 = 60
+IEC104_COT_PRIVATE_61 = 61
+IEC104_COT_PRIVATE_62 = 62
+IEC104_COT_PRIVATE_63 = 63
```

Comment:
To be honest, I've let this PR down because of those variables. I really don't like having tons of static defined variables, that could be generated automatically...

Especially, the `_fields.py` file is un-readable because of the mountain of variables :/

Reply:
We already talked about that topic while ago - https://github.com/secdev/scapy/pull/1241#discussion_r174627810.

I would like to keep all statics that are not \*\_RESERVED\_\* or \*\_PRIVATE\_\* so any ~stupid~ IDE can offer auto-complete. But I agree - those lengthy mumbo-jumbo-plus-counting-number definitions are ... pain. I already cleaned up some definitions - please check b003075 and 4142b27. If the way this goes is okay / solves the issue just drop me a line and I will happily clean up the remaining definitions.